docs/upgrade/4.x.x.adoc

```diff
@@ -0,0 +1,475 @@
+= React4XP - Upgrade - 4.x.x
+:toc: right
+
+== What
+
+=== Graal.JS
+
+We have dropped support for Nashorn in favour of Graal.JS.
+
+This means we had to upgrade the Enonic XP version requirement to 7.12
+
+Since the Application Configuration setting `react4xp.ssr.settings` was only relevant for Nashorn, it has been removed.
+
+And since React4xp now only supports Graal.JS: the Application Configuration setting `react4xp.ssr.engineName` is no longer relevant.
+
+=== TypeScript
+
+React4xp version 4 introduces support for TypeScript
+
+* https://www.npmjs.com/package/@enonic/react4xp/v/4.0.0-B1
+* https://repo.enonic.com/service/rest/repository/browse/public/com/enonic/lib/lib-react4xp/4.0.0-B1
+* https://www.npmjs.com/package/@enonic/react-components/v/3.1.2
+* https://www.npmjs.com/package/@enonic-types/lib-react4xp/v/4.0.0-B2
+
+=== React 18
+
+React4xp will now use the new React 18 functions like createRoot() and hydrateRoot() if they are available.
+
+If they are not, it will use the React 17 functions like render() and hydrate().
+
+=== Target folder (browser cache)
+
+We have changed target folder from `build/resources/main/assets/react4xp` to
+`build/resources/main/r4xAssets`. The React4xp assets are currently provided by 
+a service on top of lib-static and should not be available via assetUrl. When
+building in production mode: the asset filenames contain contenthashes, so they
+can be served as static assets with "infinite" browser caching. When building in
+development mode, React4xp uses ETag.
+
+We have not changed the source folder `src/main/resources/react4xp`, so you
+don't have to move any files.
+
+=== Named exports
+
+The default export from '/lib/enonic/react4xp' used to be a class.
+Now we are using named exports.
+You can still import the class exported under the name React4xp.
+
+=== clientRender (hydrate & ssr)
+
+The clientRender option/parameter has been removed in favour of hydrate and ssr.
+
+Read more under <<Application configuration>> and <<React4xp components>>
+
+=== Globals vs Externals
```

Comment:
I think this explanation is really good and easy to understand, so I'd not change it, but also summarize it in the end  with a table , something like:

| | Globals  | Externals |
| ------------- |------------- | ------------- |
| What is it? | ... | ... |
| Where is it used? | ... |  ...  |
| How it works? | ... | ... |

Reply:
Added